lib/controllers/issueish-detail-controller.js

```diff
@@ -44,14 +68,112 @@ export class BareIssueishDetailController extends React.Component {
+    }
+
+    // Determine if we already have this PR checked out.
+
+    const headPush = this.props.branches.getHeadBranch().getPush();
+    const headRemote = this.props.remotes.withName(headPush.getRemoteName());
+
+    // (detect checkout from pull/### refspec)
+    const fromPullRefspec =
+      headRemote.getOwner() === repository.owner.login &&
+      headRemote.getRepo() === repository.name &&
+      headPush.getShortRemoteRef() === `pulls/${issueish.number}/head`;
+
+    // (detect checkout from head repository)
+    const fromHeadRepo =
+      headRemote.getOwner() === issueish.headRepository.owner.login &&
+      headRemote.getRepo() === issueish.headRepository.name &&
+      headPush.getShortRemoteRef() === issueish.headRefName;
+
+    if (fromPullRefspec || fromHeadRepo) {
+      return this.checkoutOp.disable(checkoutStates.CURRENT, 'Current');
+    }
+
+    return this.checkoutOp.enable();
+  }
+
+  async checkout() {
+    const {repository} = this.props;
+    const {issueish} = repository;
+    const {headRepository} = issueish;
+
+    const fullHeadRef = `refs/heads/${issueish.headRefName}`;
+
+    let sourceRemoteName, localRefName;
+
+    // Discover or create a remote pointing to the repo containing the pull request's head ref.
+    // If the local repository already has the head repository specified as a remote, that remote will be used, so
+    // that any related configuration is picked up for the fetch. Otherwise, the head repository fetch URL is used
+    // directly.
+    const headRemotes = this.props.remotes.matchingGitHubRepository(headRepository.owner.login, headRepository.name);
+    if (headRemotes.length > 0) {
+      sourceRemoteName = headRemotes[0].getName();
+    } else {
+      const url = {
+        https: headRepository.url + '.git',
+        ssh: headRepository.sshUrl,
+      }[this.props.remotes.mostUsedProtocol(['https', 'ssh'])];
+
+      // This will throw if a remote with this name already exists (and points somewhere else, or we would have found
+      // it above). ¯\_(ツ)_/¯
+      const remote = await this.props.addRemote(headRepository.owner.login, url);
+      sourceRemoteName = remote.getName();
+    }
+
+    // Identify an existing local ref that already corresponds to the pull request, if one exists. Otherwise, generate
+    // a new local ref name.
+    const pullTargets = this.props.branches.getPullTargets(sourceRemoteName, fullHeadRef);
+    if (pullTargets.length > 0) {
+      localRefName = pullTargets[0].getName();
+
+      // Check out the existing local ref.
+      await this.props.checkout(localRefName);
+      await this.props.pull(fullHeadRef, {remoteName: sourceRemoteName, ffOnly: true});
+
+      return;
+    }
+
+    await this.props.fetch(fullHeadRef, {remoteName: sourceRemoteName});
+
+    // Check out the local ref and set it up to track the head ref.
+    await this.props.checkout(`pr-${issueish.number}/${issueish.headRefName}`, {
```

Comment:
GitHub Desktop creates a branch name like `pr/${issueish.number}`. Should we do the same?

Would be a bit shorter. And also avoid showing the word "master" (like `pr-123/master`) in the status-bar when forks make changes on master and then submit a PR to the upstream repo.

Reply:
Interesting. I appended the head ref name mostly so that you'd get to see the entertaining branch names that your teammates come up with :wink:

Reply:
> to see the entertaining branch names that your teammates come up with

:smile:

Ok, after more testing in Desktop, it seems that:

- If the PR is _not_ coming from a fork, the local branch name is: `${issueish.headRefName}`
- If the PR _is_ coming from a fork, the local branch name is: `pr/${issueish.number}`

The command line instructions on .com:

- If the PR is _not_ coming from a fork, the local branch name is: `${issueish.headRefName}`
- If the PR _is_ coming from a fork, the local branch name is: `${username}-${issueish.headRefName}`

Hmm.. Not sure what's the best naming scheme, but interesting to see all the differences.